File: others/src/main/java/app/lawnchair/lawnicons/helper/SvgFilesProcessor.java
package app.lawnchair.lawnicons.helper;

import com.android.ide.common.vectordrawable.Svg2Vector;
import com.google.common.base.Charsets;
import org.dom4j.Attribute;
import org.dom4j.Document;
import org.dom4j.DocumentException;
import org.dom4j.Element;
import org.dom4j.io.OutputFormat;
import org.dom4j.io.XMLWriter;

import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.EnumSet;

import static java.nio.file.FileVisitResult.CONTINUE;

public class SvgFilesProcessor {

	private final Path sourceSvgPath;
	private final Path destinationVectorPath;
	private final String extension;
	private final String extensionSuffix;
    private final CommonUtil commonUtil = new CommonUtil();
    private final String mode;


    public SvgFilesProcessor(String sourceSvgDirectory, String destinationVectorDirectory, String mode) {
        this.sourceSvgPath = Paths.get(sourceSvgDirectory);
        this.destinationVectorPath = Paths.get(destinationVectorDirectory);
        this.mode = mode;
        extensionSuffix = "";
        extension = "xml";
    }

    public void process(){
		try{
			EnumSet<FileVisitOption> options = EnumSet.of(FileVisitOption.FOLLOW_LINKS);
			//check first if source is a directory
			if(Files.isDirectory(sourceSvgPath)){
				Files.walkFileTree(sourceSvgPath, options, Integer.MAX_VALUE, new FileVisitor<>() {

					public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
						return FileVisitResult.CONTINUE;
					}

					public FileVisitResult preVisitDirectory(Path dir,
							BasicFileAttributes attrs)  {
						// Skip folder which is processing svgs to xml
						if(dir.equals(destinationVectorPath)){
							return FileVisitResult.SKIP_SUBTREE;
						}
                        Path newDirectory = destinationVectorPath.resolve(sourceSvgPath.relativize(dir));
						try{
                            Files.createDirectories(newDirectory);
						} catch(FileAlreadyExistsException ex){
                            ex.printStackTrace();
							System.out.println("FileAlreadyExistsException "+ ex);
						} catch(IOException x){
                            x.printStackTrace();
                            System.out.println("SKIP_SUBTREE  "+ x);
							return FileVisitResult.SKIP_SUBTREE;
						}
						return CONTINUE;
					}


					public FileVisitResult visitFile(Path file,
							BasicFileAttributes attrs) throws IOException {
						convertToVector(file, destinationVectorPath.resolve(sourceSvgPath.relativize(file)));
						return CONTINUE;
					}


					public FileVisitResult visitFileFailed(Path file,
							IOException exc) {
						return CONTINUE;
					}
				});
			} else {
				System.out.println("source not a directory");
			}

		} catch (IOException e){
            e.printStackTrace();
			System.out.println("IOException "+e.getMessage());
		}
	}

	private void convertToVector(Path source, Path target) throws IOException{
		// convert only if it is .svg
		if(source.getFileName().toString().endsWith(".svg")){
			String targetFile =commonUtil.getFileWithXMlExtension(target, extension, extensionSuffix);
            if(targetFile==null) {
                return;
            }
            FileOutputStream fileOutputStream = new FileOutputStream(targetFile);

            Svg2Vector.parseSvgToXml(source.toFile(), fileOutputStream);
            try {
                if(mode.equals("dark")) {
                    updatePath(targetFile, "android:strokeColor", "#000");
                    updatePath(targetFile, "android:fillColor", "#000");
                } else if (mode.equals("light")) {
                    updatePath(targetFile,"android:strokeColor","#fff");
                    updatePath(targetFile, "android:fillColor", "#fff");
                }
            } catch (DocumentException e) {
                throw new RuntimeException(e);
            }
		} else {
			System.out.println("Skipping file as its not svg "+ source.getFileName());
		}
    }

    private void updatePath(String xmlPath, String key, String value) throws DocumentException, IOException {
        Document aDocument = commonUtil.getDocument(xmlPath);
        String keyWithoutNameSpace = key.substring(key.indexOf(":")+1);
        if(aDocument!=null && aDocument.getRootElement()!=null) {
            for (Element e : aDocument.getRootElement().elements("path")) {
                Attribute attr = e.attribute(keyWithoutNameSpace);
                if (attr != null) {
                    if (!attr.getValue().equals("#00000000")) {
                        attr.setValue(String.valueOf(value));
                    }
                }
            }
            updateDocumentToFile(aDocument, xmlPath);
        }
    }
    private void updateDocumentToFile(Document outDocument, String outputConfigPath) throws IOException {
        FileWriter fileWriter = new FileWriter(outputConfigPath);
        OutputFormat format = OutputFormat.createPrettyPrint();
        format.setEncoding(Charsets.UTF_8.name());
        XMLWriter writer = new XMLWriter(fileWriter, format);
        writer.write(outDocument);
        writer.close();
    }


}

File: others/src/main/kotlin/app/lawnchair/lawnicons/helper/SvgFilesProcessor.kt
package app.lawnchair.lawnicons.helper

import com.android.ide.common.vectordrawable.Svg2Vector
import com.google.common.base.Charsets
import java.io.FileOutputStream
import java.io.FileWriter
import java.io.IOException
import java.nio.file.FileVisitOption
import java.nio.file.FileVisitResult
import java.nio.file.FileVisitor
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.attribute.BasicFileAttributes
import java.util.EnumSet
import org.dom4j.Document
import org.dom4j.DocumentException
import org.dom4j.io.OutputFormat
import org.dom4j.io.XMLWriter

class SvgFilesProcessor {
    private var sourceSvgPath: Path? = null
    private var destinationVectorPath: Path? = null
    private val extension: String? = "xml"
    private val extensionSuffix: String? = ""
    private var mode: String? = "dark"
    private val xmlUtil = XmlUtil()

    fun process(sourceDirectory: String?, destDirectory: String?, mode: String?) {
        this.sourceSvgPath = Paths.get(sourceDirectory)
        this.destinationVectorPath = Paths.get(destDirectory)
        this.mode = mode
        try {
            val options = EnumSet.of(FileVisitOption.FOLLOW_LINKS)
            // check first if source is a directory
            if (Files.isDirectory(sourceSvgPath)) {
                Files.walkFileTree(sourceSvgPath, options, Int.MAX_VALUE, fileVisitor())
            } else {
                println("source not a directory")
            }
        } catch (e: IOException) {
            println("IOException " + e.message)
        }
    }

    private fun fileVisitor() = object : FileVisitor<Path> {
        override fun postVisitDirectory(
            dir: Path?,
            exc: IOException?,
        ): FileVisitResult {
            return FileVisitResult.CONTINUE
        }

        override fun preVisitDirectory(
            dir: Path?,
            attrs: BasicFileAttributes?,
        ): FileVisitResult {
            // Skip folder which is processing svgs to xml
            if (dir == destinationVectorPath) {
                return FileVisitResult.SKIP_SUBTREE
            }
            val newDirectory = destinationVectorPath!!.resolve(
                sourceSvgPath!!.relativize(dir),
            )
            try {
                Files.createDirectories(newDirectory)
            } catch (ex: FileAlreadyExistsException) {
                ex.printStackTrace()
                println("FileAlreadyExistsException $ex")
            } catch (x: IOException) {
                x.printStackTrace()
                println("SKIP_SUBTREE  $x")
                return FileVisitResult.SKIP_SUBTREE
            }
            return FileVisitResult.CONTINUE
        }

        @Throws(IOException::class)
        override fun visitFile(
            file: Path?,
            attrs: BasicFileAttributes?,
        ): FileVisitResult {
            if (file != null) {
                convertToVector(
                    file,
                    destinationVectorPath!!.resolve(sourceSvgPath!!.relativize(file)),
                )
            }
            return FileVisitResult.CONTINUE
        }

        override fun visitFileFailed(
            file: Path,
            exc: IOException,
        ): FileVisitResult {
            return FileVisitResult.CONTINUE
        }
    }

    @Throws(IOException::class)
    private fun convertToVector(svgSource: Path, vectorTargetPath: Path) {
        // convert only if it is .svg
        if (svgSource.fileName.toString().endsWith(".svg")) {
            val targetFile =
                xmlUtil.getFileWithXMlExtension(vectorTargetPath, extension, extensionSuffix)
                    ?: return
            val fileOutputStream = FileOutputStream(targetFile)
            Svg2Vector.parseSvgToXml(svgSource.toFile(), fileOutputStream)
            try {
                if (mode == "dark") {
                    updateXmlPath(targetFile, "android:strokeColor", "#000")
                    updateXmlPath(targetFile, "android:fillColor", "#000")
                } else if (mode == "light") {
                    updateXmlPath(targetFile, "android:strokeColor", "#fff")
                    updateXmlPath(targetFile, "android:fillColor", "#fff")
                }
            } catch (e: DocumentException) {
                throw RuntimeException(e)
            }
        } else {
            println("Skipping file as its not svg " + svgSource.fileName)
        }
    }

    @Throws(DocumentException::class, IOException::class)
    private fun updateXmlPath(xmlPath: String, searchKey: String, attributeValue: String) {
        val xmlDocument = xmlUtil.getDocument(xmlPath)
        val keyWithoutNameSpace = searchKey.substring(searchKey.indexOf(":") + 1)
        if (xmlDocument != null && xmlDocument.rootElement != null) {
            for (e in xmlDocument.rootElement.elements("path")) {
                val attr = e.attribute(keyWithoutNameSpace)
                if (attr != null) {
                    if (attr.value != "#00000000") {
                        attr.value = attributeValue
                    }
                }
            }
            updateDocumentToFile(xmlDocument, xmlPath)
        }
    }

    @Throws(IOException::class)
    private fun updateDocumentToFile(outDocument: Document, outputConfigPath: String) {
        val fileWriter = FileWriter(outputConfigPath)
        val format = OutputFormat.createPrettyPrint()
        format.encoding = Charsets.UTF_8.name()
        val writer = XMLWriter(fileWriter, format)
        writer.write(outDocument)
        writer.close()
    }
}
